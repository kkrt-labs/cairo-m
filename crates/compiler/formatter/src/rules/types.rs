use cairo_m_compiler_parser::parser::{NamedType, TypeExpr};

use crate::Format;
use crate::context::FormatterCtx;
use crate::doc::Doc;
use crate::utils::*;

impl Format for TypeExpr {
    fn format(&self, ctx: &mut FormatterCtx) -> Doc {
        match self {
            Self::Named(named) => named.value().format(ctx),
            Self::Pointer(inner) => Doc::concat(vec![inner.value().format(ctx), Doc::text("*")]),
            Self::Tuple(types) => {
                let type_docs = types
                    .iter()
                    .map(|t| t.value().format(ctx))
                    .collect::<Vec<_>>();
                parens(comma_separated(type_docs))
            }
        }
    }
}

impl Format for NamedType {
    fn format(&self, _ctx: &mut FormatterCtx) -> Doc {
        match self {
            Self::Felt => Doc::text("felt"),
            Self::Bool => Doc::text("bool"),
            Self::U32 => Doc::text("u32"),
            Self::Custom(name) => Doc::text(name),
        }
    }
}
