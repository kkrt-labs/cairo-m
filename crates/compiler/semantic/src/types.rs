//! # Type System for Cairo-M
//!
//! This module implements the core type system using Salsa for incremental computation.
//! Types are represented as interned IDs for efficient comparison and storage.
//!
//! ## Architecture
//!
//! - `TypeId`: Main type identifier that wraps `TypeData`
//! - `TypeData`: The actual type information (primitives, structs, functions, etc.)
//! - `StructTypeId`: Interned struct type with fields
//! - `FunctionSignatureId`: Interned function signature with parameters and return type
//!
//! ## Design Notes
//!
//! All complex types (structs, functions) are interned separately to enable efficient
//! structural comparison and avoid deep recursion during type checking.

#[cfg(test)]
mod types_test;

use crate::place::FileScopeId;
use crate::semantic_index::DefinitionId;
use crate::SemanticDb;

/// Main type identifier that represents any type in the system
///
/// This is a Salsa-interned type that wraps `TypeData`. The interning ensures
/// that identical types have the same `TypeId`, enabling fast equality comparisons
/// and reducing memory usage.
#[salsa::interned(debug)]
pub struct TypeId<'db> {
    #[return_ref]
    pub data: TypeData<'db>,
}

/// The actual type information
///
/// This enum represents all the different kinds of types in Cairo-M.
/// Complex types like structs and functions are represented by their own
/// interned IDs to keep this enum lightweight.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum TypeData<'db> {
    /// The `felt` primitive type - Cairo's basic field element type
    Felt,

    /// A struct type, identified by its interned struct type ID
    Struct(StructTypeId<'db>),

    /// A tuple type containing an ordered list of component types
    Tuple(Vec<TypeId<'db>>),

    /// A pointer type pointing to another type
    Pointer(TypeId<'db>),

    /// A function type with its signature
    Function(FunctionSignatureId<'db>),

    /// Represents an unknown type during inference
    /// Used when type information is not yet available or during inference cycles
    Unknown,

    /// Represents a type error
    /// Used to prevent cascading errors when type checking fails
    Error,
}

/// Interned struct type definition
///
/// This contains all the semantic information about a struct type,
/// including its fields and their types. The interning ensures that
/// structurally identical struct types are represented by the same ID.
#[salsa::interned(debug)]
pub struct StructTypeId<'db> {
    /// The definition ID that corresponds to this struct in the semantic index
    pub definition_id: DefinitionId<'db>,

    /// The name of the struct
    pub name: String,

    /// The fields of the struct as an ordered map of name to type.
    /// Preserves field declaration order while allowing fast name lookup.
    /// TODO: Move to a hashmap once we figure out how to make salsa work with it.
    #[return_ref]
    pub fields: Vec<(String, TypeId<'db>)>,

    /// The scope where this struct is defined
    pub scope_id: FileScopeId,
}

/// Interned function signature
///
/// This represents the type signature of a function, including its
/// parameters and return type. Interning allows efficient comparison
/// of function types and enables function type compatibility checking.
#[salsa::interned(debug)]
pub struct FunctionSignatureId<'db> {
    /// The definition ID that corresponds to this function in the semantic index
    pub definition_id: DefinitionId<'db>,

    /// The parameters of the function as an ordered map of name to type.
    /// Preserves parameter declaration order while allowing fast name lookup.
    /// TODO: Move to a hashmap once we figure out how to make salsa work with it.
    #[return_ref]
    pub params: Vec<(String, TypeId<'db>)>,

    /// The return type of the function
    pub return_type: TypeId<'db>,
}

impl<'db> TypeData<'db> {
    /// Check if this type is a primitive type
    pub const fn is_primitive(&self) -> bool {
        matches!(self, TypeData::Felt)
    }

    /// Check if this type represents an error state
    pub const fn is_error(&self) -> bool {
        matches!(self, TypeData::Error)
    }

    /// Check if this type is unknown/unresolved
    pub const fn is_unknown(&self) -> bool {
        matches!(self, TypeData::Unknown)
    }

    /// Check if this type is a concrete type (not error or unknown)
    pub const fn is_concrete(&self) -> bool {
        !self.is_error() && !self.is_unknown()
    }

    /// Get a human-readable display name for this type
    pub fn display_name(&self, db: &dyn SemanticDb) -> String {
        match self {
            TypeData::Felt => "felt".to_string(),
            TypeData::Struct(struct_id) => struct_id.name(db),
            TypeData::Tuple(types) => {
                let type_names: Vec<String> =
                    types.iter().map(|t| t.data(db).display_name(db)).collect();
                format!("({})", type_names.join(", "))
            }
            TypeData::Pointer(inner) => {
                format!("{}*", inner.data(db).display_name(db))
            }
            TypeData::Function(_) => "function".to_string(),
            TypeData::Unknown => "<unknown>".to_string(),
            TypeData::Error => "<error>".to_string(),
        }
    }
}

impl<'db> StructTypeId<'db> {
    /// Get the type of a specific field by name
    pub fn field_type(&self, db: &'db dyn SemanticDb, field_name: &str) -> Option<TypeId<'db>> {
        self.fields(db)
            .iter()
            .find(|(name, _)| name == field_name)
            .map(|(_, type_id)| *type_id)
    }

    /// Check if this struct has a field with the given name
    pub fn has_field(&self, db: &dyn SemanticDb, field_name: &str) -> bool {
        self.fields(db).iter().any(|(name, _)| name == field_name)
    }

    /// Get all field names in declaration order
    pub fn field_names(&self, db: &'db dyn SemanticDb) -> Vec<String> {
        self.fields(db)
            .iter()
            .map(|(name, _)| name.clone())
            .collect()
    }
}

impl<'db> FunctionSignatureId<'db> {
    /// Get the type of a specific parameter by name
    pub fn param_type(&self, db: &'db dyn SemanticDb, param_name: &str) -> Option<TypeId<'db>> {
        self.params(db)
            .iter()
            .find(|(name, _)| name == param_name)
            .map(|(_, type_id)| *type_id)
    }

    /// Check if this function has a parameter with the given name
    pub fn has_param(&self, db: &dyn SemanticDb, param_name: &str) -> bool {
        self.params(db).iter().any(|(name, _)| name == param_name)
    }

    /// Get all parameter names in declaration order
    pub fn param_names(&self, db: &'db dyn SemanticDb) -> Vec<String> {
        self.params(db)
            .iter()
            .map(|(name, _)| name.clone())
            .collect()
    }

    /// Get the number of parameters
    pub fn param_count(&self, db: &dyn SemanticDb) -> usize {
        self.params(db).len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::db::tests::test_db;

    #[test]
    fn test_type_data_classification() {
        let felt_data = TypeData::Felt;
        assert!(felt_data.is_primitive());
        assert!(felt_data.is_concrete());
        assert!(!felt_data.is_error());
        assert!(!felt_data.is_unknown());

        let error_data = TypeData::Error;
        assert!(error_data.is_error());
        assert!(!error_data.is_concrete());

        let unknown_data = TypeData::Unknown;
        assert!(unknown_data.is_unknown());
        assert!(!unknown_data.is_concrete());
    }

    #[test]
    fn test_type_display_names() {
        let db = test_db();
        assert_eq!(TypeData::Felt.display_name(&db), "felt");
        assert_eq!(TypeData::Error.display_name(&db), "<error>");
        assert_eq!(TypeData::Unknown.display_name(&db), "<unknown>");
    }

    #[test]
    fn test_type_interning() {
        let db = test_db();

        // Same TypeData should produce the same TypeId
        let type1 = TypeId::new(&db, TypeData::Felt);
        let type2 = TypeId::new(&db, TypeData::Felt);
        assert_eq!(type1, type2);

        // Different TypeData should produce different TypeId
        let type3 = TypeId::new(&db, TypeData::Error);
        assert_ne!(type1, type3);
    }
}
