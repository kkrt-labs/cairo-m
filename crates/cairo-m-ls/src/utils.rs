//! Utility functions shared across the language server

use std::path::PathBuf;

use tower_lsp::lsp_types::Url;
use tracing::debug;

/// Convert path string to URI
///
/// Handles both regular file paths and URI strings starting with "file://"
pub fn get_uri_from_path_str(path_str: &str) -> Result<Url, String> {
    if path_str.starts_with("file://") {
        Url::parse(path_str).map_err(|e| format!("Failed to parse URI: {}", e))
    } else {
        Url::from_file_path(path_str)
            .map_err(|_| format!("Failed to convert path to URI: {}", path_str))
    }
}

/// Get PathBuf from diagnostic file path and URI
///
/// Handles conversion between URI and file path formats
pub fn get_path_from_diagnostic(uri: &Url, file_path: &str) -> Option<PathBuf> {
    if file_path.starts_with("file://") {
        match uri.to_file_path() {
            Ok(path) => Some(path),
            Err(_) => {
                debug!("Warning: Failed to convert URI to file path: {}", uri);
                None
            }
        }
    } else {
        Some(PathBuf::from(file_path))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_get_uri_from_path_str_regular_path() {
        let path = "/tmp/test.cm";
        let uri = get_uri_from_path_str(path).unwrap();
        assert_eq!(uri.scheme(), "file");
        assert!(uri.path().ends_with("/tmp/test.cm"));
    }

    #[test]
    fn test_get_uri_from_path_str_uri_format() {
        let uri_str = "file:///tmp/test.cm";
        let uri = get_uri_from_path_str(uri_str).unwrap();
        assert_eq!(uri.scheme(), "file");
        assert_eq!(uri.path(), "/tmp/test.cm");
    }

    #[test]
    fn test_get_path_from_diagnostic_uri_format() {
        let uri = Url::parse("file:///tmp/test.cm").unwrap();
        let file_path = "file:///tmp/test.cm";
        let path = get_path_from_diagnostic(&uri, file_path).unwrap();
        assert_eq!(path, PathBuf::from("/tmp/test.cm"));
    }

    #[test]
    fn test_get_path_from_diagnostic_regular_path() {
        let uri = Url::parse("file:///tmp/test.cm").unwrap();
        let file_path = "/tmp/test.cm";
        let path = get_path_from_diagnostic(&uri, file_path).unwrap();
        assert_eq!(path, PathBuf::from("/tmp/test.cm"));
    }
}
