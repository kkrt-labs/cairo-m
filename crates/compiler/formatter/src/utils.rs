use crate::doc::Doc;

/// Helper function to create a space-separated list
pub fn space_separated(items: Vec<Doc>) -> Doc {
    Doc::join(Doc::text(" "), items)
}

/// Helper function to create a comma-separated list
pub fn comma_separated(items: Vec<Doc>) -> Doc {
    Doc::join(Doc::text(", "), items)
}

/// Helper function to wrap in parentheses
pub fn parens(inner: Doc) -> Doc {
    Doc::concat(vec![Doc::text("("), inner, Doc::text(")")])
}

/// Helper function to wrap in braces
pub fn braces(inner: Doc) -> Doc {
    Doc::concat(vec![Doc::text("{"), inner, Doc::text("}")])
}

/// Helper function to wrap in brackets
pub fn brackets(inner: Doc) -> Doc {
    Doc::concat(vec![Doc::text("["), inner, Doc::text("]")])
}
