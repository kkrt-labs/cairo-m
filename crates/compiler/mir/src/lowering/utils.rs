//! # Lowering Utilities
//!
//! This module contains shared utility functions used across the lowering
//! implementation.

use cairo_m_compiler_parser::parser::{Expression, Spanned};
use cairo_m_compiler_semantic::definition::DefinitionKind;
use cairo_m_compiler_semantic::place::FileScopeId;
use cairo_m_compiler_semantic::semantic_index::{DefinitionId, DefinitionIndex, ExpressionId};
use cairo_m_compiler_semantic::type_resolution::{
    definition_semantic_type, expression_semantic_type,
};
use cairo_m_compiler_semantic::types::TypeData;
use cairo_m_compiler_semantic::SemanticIndex;

use crate::instruction::CalleeSignature;
use crate::mir_types::InstructionEmitter;
use crate::{BasicBlockId, FunctionId, Instruction, MirType, Value, ValueId};

use super::builder::MirBuilder;

impl<'a, 'db> MirBuilder<'a, 'db> {
    /// Gets the MIR type for an expression by its ID
    pub fn get_expression_type(&self, expr_id: ExpressionId) -> MirType {
        let semantic_type =
            expression_semantic_type(self.ctx.db, self.ctx.crate_id, self.ctx.file, expr_id, None);
        MirType::from_semantic_type(self.ctx.db, semantic_type)
    }

    /// Checks if we're currently in a loop context
    pub const fn in_loop(&self) -> bool {
        !self.state.loop_stack.is_empty()
    }

    /// Gets the current loop's continue and break targets
    pub fn current_loop_targets(&self) -> Option<(BasicBlockId, BasicBlockId)> {
        self.state.loop_stack.last().copied()
    }

    /// Binds a value to a variable identifier with complete lifecycle management
    ///
    /// This helper encapsulates:
    /// 1. Resolving the identifier to its DefinitionId and MirDefinitionId
    /// 2. Checking if the variable is used (and early return if not)
    /// 3. For simple values (not addresses), directly mapping the variable to the value
    /// 4. For addresses, allocating stack space and storing the value
    /// 5. Updating the definition_to_value mapping
    pub fn bind_variable(
        &mut self,
        name: &Spanned<String>,
        scope: FileScopeId,
        value: Value,
    ) -> Result<(), String> {
        // Resolve the identifier to its definition
        let (def_idx, _) = self
            .ctx
            .semantic_index
            .resolve_name_to_definition(name.value(), scope)
            .ok_or_else(|| {
                format!(
                    "Failed to resolve variable '{}' in scope {:?}",
                    name.value(),
                    scope
                )
            })?;

        let def_id = DefinitionId::new(self.ctx.db, self.ctx.file, def_idx);
        let mir_def_id = self.convert_definition_id(def_id);
        let is_used = is_definition_used(self.ctx.semantic_index, def_idx);
        if !is_used {
            // Map to a dummy value and return early
            let dummy_addr = self.state.mir_function.new_value_id();
            self.state
                .definition_to_value
                .insert(mir_def_id, dummy_addr);
            return Ok(());
        }

        // Check if we can use the value directly without allocation
        // This optimization avoids unnecessary stack allocation for simple values
        match value {
            Value::Operand(value_id) => {
                // Check if this operand is already a non-pointer value we can use directly
                if let Some(value_type) = self.state.mir_function.get_value_type(value_id) {
                    if !matches!(value_type, MirType::Pointer(_)) {
                        // It's a value (not a pointer) - we can use it directly without allocation!
                        self.state.definition_to_value.insert(mir_def_id, value_id);
                        return Ok(());
                    }
                }
            }
            Value::Literal(_) => {
                // Literals are immediate values - create a value instruction for them
                let semantic_type =
                    definition_semantic_type(self.ctx.db, self.ctx.crate_id, def_id);
                let var_type = MirType::from_semantic_type(self.ctx.db, semantic_type);
                let value_id = self.state.mir_function.new_typed_value_id(var_type.clone());

                // Create a move/immediate instruction
                self.instr()
                    .add_instruction(var_type.emit_assign(value_id, value));

                // Map the variable directly to this value
                self.state.definition_to_value.insert(mir_def_id, value_id);
                return Ok(());
            }
            _ => {
                // For other cases, fall through to allocation
            }
        }

        // For addresses or complex values, we need to allocate and store
        // Get the variable's semantic type and convert to MirType
        let semantic_type = definition_semantic_type(self.ctx.db, self.ctx.crate_id, def_id);
        let var_type = MirType::from_semantic_type(self.ctx.db, semantic_type);

        // Allocate space for the variable
        let var_addr = self
            .state
            .mir_function
            .new_typed_value_id(MirType::pointer(var_type.clone()));

        let mut instr = self.instr();
        instr.add_instruction(Instruction::stack_alloc(var_addr, var_type.size_units()));

        // Emit the appropriate store instruction based on type
        instr.add_instruction(var_type.emit_store(Value::operand(var_addr), value));

        // Update the definition to value mapping
        self.state.definition_to_value.insert(mir_def_id, var_addr);
        Ok(())
    }

    /// Resolves a function call's callee to its FunctionId
    ///
    /// This centralizes the logic for resolving function calls, whether they're
    /// local functions or imported ones. Handles both simple identifiers and
    /// member access patterns.
    pub fn resolve_function(&mut self, callee: &Spanned<Expression>) -> Result<FunctionId, String> {
        match callee.value() {
            Expression::Identifier(func_name) => {
                // Get the scope for the callee from its expression info
                let callee_expr_id = self
                    .ctx
                    .semantic_index
                    .expression_id_by_span(callee.span())
                    .ok_or_else(|| {
                        format!("No ExpressionId found for callee span {:?}", callee.span())
                    })?;
                let callee_expr_info = self
                    .ctx
                    .semantic_index
                    .expression(callee_expr_id)
                    .ok_or_else(|| format!("No ExpressionInfo for callee ID {callee_expr_id:?}"))?;

                // Resolve the function name in the appropriate scope
                let (local_def_idx, local_def) = self
                    .ctx
                    .semantic_index
                    .resolve_name_to_definition(func_name.value(), callee_expr_info.scope_id)
                    .ok_or_else(|| {
                        format!(
                            "Failed to resolve function '{}' in scope {:?}",
                            func_name.value(),
                            callee_expr_info.scope_id
                        )
                    })?;

                // Handle function resolution: local functions vs imported functions
                match &local_def.kind {
                    DefinitionKind::Function(_) => {
                        // Local function - use current file
                        let func_def_id =
                            DefinitionId::new(self.ctx.db, self.ctx.file, local_def_idx);
                        if let Some((_, func_id)) = self.ctx.function_mapping.get(&func_def_id) {
                            Ok(*func_id)
                        } else {
                            Err(format!(
                                "Function '{}' not found in function mapping",
                                func_name.value()
                            ))
                        }
                    }
                    DefinitionKind::Use(use_ref) => {
                        // Imported function - follow the import chain
                        self.resolve_imported_function(
                            use_ref.imported_module.value(),
                            func_name.value(),
                        )
                        .ok_or_else(|| {
                            format!(
                                "Failed to resolve imported function '{}' from module '{}'",
                                func_name.value(),
                                use_ref.imported_module.value()
                            )
                        })
                    }
                    _ => Err(format!(
                        "'{}' is not a function or import",
                        func_name.value()
                    )),
                }
            }
            Expression::MemberAccess { .. } => {
                // For member access patterns (e.g., module.function), use existing resolution
                self.resolve_callee_expression(callee)
            }
            _ => Err("Unsupported callee expression type".to_string()),
        }
    }

    /// Emits a call instruction with destinations and proper signature
    ///
    /// This helper centralizes the logic for emitting function calls with
    /// proper signatures and destination handling.
    pub fn emit_call_with_destinations(
        &mut self,
        func_id: FunctionId,
        args: Vec<Value>,
        dests: Vec<ValueId>,
    ) -> Result<(), String> {
        // Get the function signature
        let (param_types, return_types) = self.get_function_signature(func_id)?;

        // Create the CalleeSignature
        let signature = CalleeSignature {
            param_types,
            return_types,
        };

        // Create the call instruction with the signature
        let call_instr = Instruction::call(dests, func_id, args, signature);
        self.instr().add_instruction(call_instr);
        Ok(())
    }

    /// Emits a call instruction and discards the result(s)
    ///
    /// This is used for function calls in expression statements where the
    /// return value is not used. It handles both void functions and functions
    /// that return values by creating dummy destinations.
    pub fn emit_call_and_discard_result(
        &mut self,
        func_id: FunctionId,
        args: Vec<Value>,
        expr_id: cairo_m_compiler_semantic::semantic_index::ExpressionId,
    ) -> Result<(), String> {
        // Check the function's return type
        let func_expr_semantic_type =
            expression_semantic_type(self.ctx.db, self.ctx.crate_id, self.ctx.file, expr_id, None);

        match func_expr_semantic_type.data(self.ctx.db) {
            TypeData::Tuple(element_types) if element_types.is_empty() => {
                // Function returns unit/void
                let (param_types, return_types) = self.get_function_signature(func_id)?;
                let signature = CalleeSignature {
                    param_types,
                    return_types,
                };
                self.instr().void_call(func_id, args, signature);
            }
            TypeData::Tuple(element_types) => {
                // Function returns a tuple - create destinations but don't use them
                let mut dests = Vec::new();
                for elem_type in element_types {
                    let mir_type = MirType::from_semantic_type(self.ctx.db, elem_type);
                    dests.push(self.state.mir_function.new_typed_value_id(mir_type));
                }
                self.emit_call_with_destinations(func_id, args, dests)?;
            }
            _ => {
                // Function returns a single value - create a destination but don't use it
                let return_type = MirType::from_semantic_type(self.ctx.db, func_expr_semantic_type);
                let dest = self.state.mir_function.new_typed_value_id(return_type);
                self.emit_call_with_destinations(func_id, args, vec![dest])?;
            }
        }
        Ok(())
    }
}

pub(crate) fn is_definition_used(semantic_index: &SemanticIndex, def_idx: DefinitionIndex) -> bool {
    semantic_index.definition(def_idx).is_none_or(|definition| {
        semantic_index
            .place_table(definition.scope_id)
            .is_none_or(|place_table| {
                place_table
                    .place(definition.place_id)
                    .is_none_or(|place| place.is_used())
            })
    })
}
