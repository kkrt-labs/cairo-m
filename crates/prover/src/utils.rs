use num_traits::{One, Zero};
use stwo_prover::core::backend::simd::conversion::Pack;
use stwo_prover::core::backend::simd::m31::{N_LANES, PackedM31};
use stwo_prover::core::fields::m31::M31;

use crate::adapter::ExecutionBundle;

/// The enabler column is a column of length `padding_offset.next_power_of_two()` where
/// 1. The first `padding_offset` elements are set to 1;
/// 2. The rest are set to 0.
#[derive(Debug, Clone)]
pub struct Enabler {
    pub padding_offset: usize,
}
impl Enabler {
    pub const fn new(padding_offset: usize) -> Self {
        Self { padding_offset }
    }

    pub fn packed_at(&self, vec_row: usize) -> PackedM31 {
        let row_offset = vec_row * N_LANES;
        if row_offset >= self.padding_offset {
            return PackedM31::zero();
        }
        if row_offset + N_LANES <= self.padding_offset {
            return PackedM31::one();
        }

        // The row is partially enabled.
        let mut res = [M31::zero(); N_LANES];
        for v in res.iter_mut().take(self.padding_offset - row_offset) {
            *v = M31::one();
        }
        PackedM31::from_array(res)
    }
}

// Flattened PackedExecutionBundle that contains all the M31 components as separate PackedM31 vectors
// This structure is optimized for SIMD operations
// Note: For operands (mem1-3), we only store single M31 values since DataAccess only has M31 fields
#[derive(Debug, Clone, Copy)]
pub struct PackedExecutionBundle {
    // VM registers (2 fields)
    pub pc: PackedM31,
    pub fp: PackedM31,

    pub clock: PackedM31,

    // Memory arg 0 - Instruction (8 fields: 4 value M31s, prev_clock)
    pub inst_prev_clock: PackedM31,
    pub inst_value_0: PackedM31,
    pub inst_value_1: PackedM31,
    pub inst_value_2: PackedM31,
    pub inst_value_3: PackedM31,

    // Memory arg 1 - Operand 0 (4 fields: address, value, prev_value, prev_clock)
    pub mem1_prev_value: PackedM31,
    pub mem1_value: PackedM31,
    pub mem1_prev_clock: PackedM31,

    // Memory arg 2 - Operand 1 (4 fields)
    pub mem2_prev_value: PackedM31,
    pub mem2_value: PackedM31,
    pub mem2_prev_clock: PackedM31,

    // Memory arg 3 - Operand 2 (4 fields)
    pub mem3_prev_value: PackedM31,
    pub mem3_value: PackedM31,
    pub mem3_prev_clock: PackedM31,
}

impl Pack for ExecutionBundle {
    type SimdType = PackedExecutionBundle;

    fn pack(inputs: [Self; N_LANES]) -> Self::SimdType {
        PackedExecutionBundle {
            // Pack VM registers
            pc: PackedM31::from_array(std::array::from_fn(|i| inputs[i].registers.pc)),
            fp: PackedM31::from_array(std::array::from_fn(|i| inputs[i].registers.fp)),

            // Pack clock
            clock: PackedM31::from_array(std::array::from_fn(|i| inputs[i].clock)),

            // Memory arg 0 - Instruction (value is QM31)
            inst_prev_clock: PackedM31::from_array(std::array::from_fn(|i| {
                inputs[i].instruction.prev_clock
            })),
            inst_value_0: PackedM31::from_array(std::array::from_fn(|i| {
                inputs[i].instruction.value.0.0
            })),
            inst_value_1: PackedM31::from_array(std::array::from_fn(|i| {
                inputs[i].instruction.value.0.1
            })),
            inst_value_2: PackedM31::from_array(std::array::from_fn(|i| {
                inputs[i].instruction.value.1.0
            })),
            inst_value_3: PackedM31::from_array(std::array::from_fn(|i| {
                inputs[i].instruction.value.1.1
            })),

            // Memory arg 1
            mem1_prev_value: PackedM31::from_array(std::array::from_fn(|i| {
                inputs[i].operands[0].map_or_else(M31::zero, |op| op.prev_value)
            })),
            mem1_value: PackedM31::from_array(std::array::from_fn(|i| {
                inputs[i].operands[0].map_or_else(M31::zero, |op| op.value)
            })),
            mem1_prev_clock: PackedM31::from_array(std::array::from_fn(|i| {
                inputs[i].operands[0].map_or_else(M31::zero, |op| op.prev_clock)
            })),

            // Memory arg 2
            mem2_prev_value: PackedM31::from_array(std::array::from_fn(|i| {
                inputs[i].operands[1].map_or_else(M31::zero, |op| op.prev_value)
            })),
            mem2_value: PackedM31::from_array(std::array::from_fn(|i| {
                inputs[i].operands[1].map_or_else(M31::zero, |op| op.value)
            })),
            mem2_prev_clock: PackedM31::from_array(std::array::from_fn(|i| {
                inputs[i].operands[1].map_or_else(M31::zero, |op| op.prev_clock)
            })),

            // Memory arg 3
            mem3_prev_value: PackedM31::from_array(std::array::from_fn(|i| {
                inputs[i].operands[2].map_or_else(M31::zero, |op| op.prev_value)
            })),
            mem3_value: PackedM31::from_array(std::array::from_fn(|i| {
                inputs[i].operands[2].map_or_else(M31::zero, |op| op.value)
            })),
            mem3_prev_clock: PackedM31::from_array(std::array::from_fn(|i| {
                inputs[i].operands[2].map_or_else(M31::zero, |op| op.prev_clock)
            })),
        }
    }
}
