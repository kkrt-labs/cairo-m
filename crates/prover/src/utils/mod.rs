pub mod enabler;
pub mod execution_bundle;
pub mod poseidon;
