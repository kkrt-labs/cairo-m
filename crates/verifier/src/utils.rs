use num_traits::{One, Zero};
use stwo_prover::core::backend::simd::m31::{PackedM31, N_LANES};
use stwo_prover::core::fields::m31::M31;

/// The enabler column is a column of length `padding_offset.next_power_of_two()` where
/// 1. The first `padding_offset` elements are set to 1;
/// 2. The rest are set to 0.
#[derive(Debug, Clone)]
pub struct Enabler {
    pub padding_offset: usize,
}
impl Enabler {
    pub const fn new(padding_offset: usize) -> Self {
        Self { padding_offset }
    }

    pub fn packed_at(&self, vec_row: usize) -> PackedM31 {
        let row_offset = vec_row * N_LANES;
        if row_offset >= self.padding_offset {
            return PackedM31::zero();
        }
        if row_offset + N_LANES <= self.padding_offset {
            return PackedM31::one();
        }

        // The row is partially enabled.
        let mut res = [M31::zero(); N_LANES];
        res[..self.padding_offset - row_offset].fill(M31::one());
        PackedM31::from_array(res)
    }
}
