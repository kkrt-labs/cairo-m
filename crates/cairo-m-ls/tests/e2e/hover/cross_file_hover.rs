use super::super::support::{Fixture, extract_cursors};
use crate::hover::HoverTransformer;
use crate::support::insta::test_transform;

#[tokio::test]
async fn test_hover_on_imported_function() {
    let fixture = Fixture::new();
    fixture.add_cairom_toml("test_project");
    fixture.add_file(
        "main.cm",
        r#"
use utils::helper_foo;

func main() {
    let result = helper_foo(42);
}
"#,
    );
    fixture.add_file(
        "utils.cm",
        r#"
func helper_foo(x: felt) -> felt {
    return x * 2;
}
"#,
    );

    let (_, cursors) = extract_cursors(
        r#"
use utils::helper_foo;

func main() {
    let result = helper_<caret>foo(42);
}
"#,
    );

    test_transform!(HoverTransformer, fixture, cursors, |result: &str| {
        assert!(result.contains("function"), "{}", result);
    });
}

#[ignore = "TODO: Type info not available for user-defined types"]
#[tokio::test]
async fn test_hover_on_imported_type() {
    let fixture = Fixture::new();
    fixture.add_cairom_toml("test_project");
    fixture.add_file(
        "main.cm",
        r#"
use types::CustomType;

func main() {
    let value: CustomType = CustomType { value: 42 };
}
"#,
    );
    fixture.add_file(
        "types.cm",
        r#"
struct CustomType {
    value: felt,
}
"#,
    );

    let (_, cursors) = extract_cursors(
        r#"
use types::CustomType;

func main() {
    let value: Custom<caret>Type = CustomType { value: 42 };
}
"#,
    );

    test_transform!(HoverTransformer, fixture, cursors, |result: &str| {
        assert!(result.contains("felt"), "{}", result);
    });
}

#[tokio::test]
async fn test_hover_on_module_name() {
    let fixture = Fixture::new();
    fixture.add_cairom_toml("test_project");
    fixture.add_file(
        "main.cm",
        r#"
use utils::calculate;

func main() {
    let result = calculate(10, 20);
}
"#,
    );
    fixture.add_file(
        "utils.cm",
        r#"
func calculate(a: felt, b: felt) -> felt {
    a + b
}

func multiply(a: felt, b: felt) -> felt {
    a * b
}
"#,
    );

    let (_, cursors) = extract_cursors(
        r#"
use util<caret>s::calculate;

func main() {
    let result = calculate(10, 20);
}
"#,
    );

    test_transform!(HoverTransformer, fixture, cursors, |result: &str| {
        assert!(result.contains("No hover info"), "{}", result);
    });
}

#[tokio::test]
async fn test_hover_on_imported_constant() {
    let fixture = Fixture::new();
    fixture.add_cairom_toml("test_project");
    fixture.add_file(
        "main.cm",
        r#"
use constants::MAX_VALUE;

func main() {
    let limit = MAX_VALUE;
}
"#,
    );
    fixture.add_file(
        "constants.cm",
        r#"
const MAX_VALUE = 1000;
"#,
    );

    let (_, cursors) = extract_cursors(
        r#"
use constants::MAX_VALUE;

func main() {
    let limit = MAX_<caret>VALUE;
}
"#,
    );

    test_transform!(HoverTransformer, fixture, cursors, |result: &str| {
        assert!(result.contains("felt"), "{}", result);
    });
}
