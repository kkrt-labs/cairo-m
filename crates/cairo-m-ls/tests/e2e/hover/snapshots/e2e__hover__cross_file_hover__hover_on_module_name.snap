---
source: crates/cairo-m-ls/tests/e2e/hover/cross_file_hover.rs
expression: result
---
No hover info
